### PR TITLE
ENH Add burn and thin kwargs to sample.

### LINKED ISSUE #1010: Specify burn-in to sample
While keeping all generated samples is fine for HMC/NUTS, it is not practical for Metropolis sampling. It is often necessary to run 100K+ iterations before convergence, and these values should not be stored.

### PR DESCRIPTION
**Pull request by twiecki**
Closes #1010.

### CODE REVIEW

**Comment by ColCarroll** [comment]
LGTM

**Comment by fonnesbeck** [comment]
There is never a good reason to thin traces.

**Comment by twiecki** [comment]
Not even to save memory?

**Comment by fonnesbeck** [comment]
If you are thinning just to save memory, then you should not have sampled as much in the first place. Almost always (in fact, I can't think of any other use), thinning is performed to deal with autocorrelation. All that autocorrelation does (in a converged chain) is reduce your effective sample size. So, thinning is doing exactly what autocorrelation is doing -- reducing your sample size. If you want to reduce autocorrelation, you need to change your model, change your sampler, or both. Thinning is of no benefit.

File: pymc3/sampling.py [hunk 1]
```diff
@@ -265,8 +276,9 @@ def _iter_sample(draws, step, start=None, trace=None, chain=0, tune=None,
         if i == tune:
             step = stop_tuning(step)
         point = step.step(point)
-        strace.record(point)
-        yield strace
+        if (i % thin == 0) and (i >= burn):
```
Comment (fonnesbeck) [inline pymc3/sampling.py:279]: I'm definitely not a fan of adding *two* condition checks at every iteration to add a feature we already had.

**Comment by twiecki** [comment]
OK let's revert the PR for now.

**Pull request merged**